Generate typeclass instances

MajaRet opened the pull request:
### Issue

Closes #150
Closes #151 

### Description of the Change

The compiler now generates `Normalform` and `ShareableArgs` instances for user-defined data types.

In order to properly handle a program, it needs to be normalized first so that handlers do not have to consider effects in the _components_ of data structures. 
To be able to perform this normalization, a data type (and its argument types) must have a `Normalform` instance.

Furthermore, the `share` operator requires that the type of the computation that is shares has a `ShareableArgs` instance so its components can be shared.

`Normalform` and `ShareableArgs` instances exist for all Prelude types, but users will also define their own Haskell data types. Therefore, the compiler needs to generate these instances for user-defined data types.


### Verification Process

Tests verifying the functionality of generated `Normalform` instances for various types have been added to the `example` folder by @marvin2706. Additional functions and tests have been added (in the same modules) to test the generated `ShareableArgs` instances.

### Additional Notes

Because the generation of induction schemes has been moved into its own module, there will likely be merge conflicts with #206.

Commit:
Implement first part of Normalform instance generator #150

During the conversion of a data declaration, a Normalform instance
will be generated.
This commit implements a part of that generator that generates
most of the nf' function. The resulting code is not yet
valid because the type signature is still missing, but if the
type signature is added manually, the generated code is valid.

We are able to consider types with nested recursion and mutually
recursive types, but I have not considered type synonyms yet.
That will also be added later.

Commit:
Generate full Normalform instances #150

Commit:
Always export Identity #150

Commit:
Clean up code a little #150

Commit:
Merge with branch issue-148#150

Commit:
Format code #150

Commit:
Merge main into issue-150 #150

Commit:
Refactor code #150

Commit:
Refactor code and expand type synonyms #150

Commit:
Generalize instance generation #150

The program to generate typeclass instances for user-defined Haskell
types is now more general.
Instances for different typeclasses can now be generated simply by
passing a few parameters, namely:
- The name of the class
- The name of the function provided by the class
- A function that generates appropriate binders and return types
- A function that builds a concrete value of the return type

Currently, only typeclass instances with a certain structure can be
generated (for example, the class can currently only contain one
function), but it should be quite easy to generate instances for
ShareableArgs in addition to Normalform now.

Commit:
Use fresh variables for unification and adjust local environments #150

Before the return type of a data constructor is unified with a type
expression, all variables (underscores) in the type expression are
replaced with fresh variables to prevent unification failures.

Additionally, the naming of the instance and top-level functions is
now done outside of a local environment so that those names are
registered globally and no name clashes can occur.
Local functions and variables are still named inside a local environment.

Commit:
Add helper functions to FreeC.IR.Syntax.Type #150

Commit:
Add smart constructor for qualified Coq names to Coq.Syntax #150

Commit:
Add some constants to Coq.Base #150

Commit:
Use helper function from IR.Syntax.Type in CompletePatternPass #150

Commit:
Refactor code a little #150

Commit:
Refactor code some more #150

Commit:
Remove redundant brackets #150

MarvinLira committed:
Add examples to test generated normalforms in Coq #175

MarvinLira committed:
Run floskell #175

Commit:
Add documentation and refactor code #150

Commit:
Merge branch 'main' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Merge partially with issue-183 #150

Commit:
Merge fully with issue-183 #150

Commit:
Refactor code and remove Shape and Pos arguments from local functions #150

Commit:
Refactor code, add documentation and fix a bug #150

The code has been restructured in a way that greatly reduces
the number of maps and zips.

Haddock documentation and regular comments have been added to the
main functions.

In addition, a bug has been fixed where unification would produce
incorrect results because type synonyms in a data constructor's
argument types were not expanded.

Commit:
Merge branch 'issue-150' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Use Map for type map #150

The map mapping types to function names is now defined using a
predefined map.
For that reason, Ord instances had to be added to Type and SrcSpan.

Commit:
Fix refactoring-induced bug #150

Commit:
Merge branch 'issue-176' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Merge branch 'issue-176' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Use Coq.Base functions added to the main branch #150

Commit:
Merge branch 'main' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Make helper functions local #150

Commit:
Expand documentation #150

Commit:
Add Coq comment above class instances #150

Commit:
Add constant for underscore variable #150

Commit:
Format code with Floskell #150

Commit:
Format Coq.Base #150

Commit:
Merge main into issue-150 #150

Commit:
Adjust tests #150

Commit:
Merge main into issue-150 #150

Commit:
Format code with Floskell #150

Commit:
Generate ShareableArgs instances #150 #151

Commit:
Adjust tests #150 #151

Commit:
Apply HLint hint #150 #151

Commit:
Add tests for generated ShareableArgs instances #150 #151

Commit:
Rename Normalform and NormalformProofs #150 #151

Commit:
Format code with Floskell #150 #151

Commit:
Fix indentation of Haddock comment #150 #151

Commit:
Format code with Floskell #150 #151

Commit:
Add a few more example type to test generated instances #150 #151

MajaRet committed:
Add periods to comments #150 #151

MarvinLira commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+--   For example, if we want to generate an instance for the Haskell type
+--   @data Forest a = AForest [Forest a]
+--                  | IntForest [Forest Int]
+--                  | BoolForest [ForestBool]@,
```
```suggestion
--                  | BoolForest [Forest Bool]@,
```

MarvinLira commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+    --   This is necessary so that, for example, @Pair Bool t@ is not
+    --   translated to @_ t@, but to @Pair _ t@.
+    stripType' :: IR.Type -> Bool -> IR.Type
+
+    --
```
```suggestion
```

MarvinLira reviewed (changes requested):
In `TypeDecl` you are defining the names of the classes `Normalform` and `ShareableArgs` as hardcoded strings. You should use the names defined in `Coq.Base` instead.

Functions that you require from the `base` library should also be defined in `Coq.Base` instead of `TypeDecl` and be added to the list of reserved identifiers. Otherwise name clashes with user defined functions could be produced. I noticed this with `nf` and `cbneed`.

I think the name clash problem does not apply to the class functions `nf'` and `shareArgs`. So there it would be up to you whether you add them too or not.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  -------------------------------------------------------------------------------
+  -------------------------------------------------------------------------------
+  -- Functions to produce Normalform instances                                 --
+  -------------------------------------------------------------------------------
```
One of the headings should be removed. Also capitalize all words except for articles, conjunctions and prepositions in headings.

```suggestion
  -------------------------------------------------------------------------------
  -- Functions to Produce Normalform Instances                                 --
  -------------------------------------------------------------------------------
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  -- The names of the constructors of the data types for which
+  -- we are defining instances.
+  conNames :: [IR.TypeConName]
+  conNames = map IR.typeDeclQName dataDecls
```
The name and comment suggest that this is a list of data constructor but this is a list of data constructors.

```suggestion
  -- The names of the type constructors of the data types for which
  -- we are defining instances.
  typeConNames :: [IR.TypeConName]
  typeConNames = map IR.typeDeclQName dataDecls
```

just95 commented on src/test/FreeC/Backend/Coq/Converter/TypeDeclTests.hs:
```diff
@@ -110,6 +113,61 @@ testConvertTypeDecl
+          ++ "Instance NormalformTree_ {Shape : Type} {Pos : Shape -> Type} "
+          ++ "{a b : Type} `{Normalform Shape Pos a b} "
+          ++ ": Normalform Shape Pos (Tree Shape Pos a) "
+          ++ "(Tree Identity.Shape Identity.Pos b) := { nf' := nf'Tree_ }. "
```
By indenting the code, the tests would become much easier to read. 

```suggestion
           ++ " (* Normalform instance for Tree *) "
           ++ "Fixpoint nf'Tree_"
           ++ "  {Shape : Type} {Pos : Shape -> Type}"
           ++ "  {a b : Type} `{Normalform Shape Pos a b}"
           ++ "  (x : Tree Shape Pos a)"
           ++ "  : Free Shape Pos (Tree Identity.Shape Identity.Pos b)"
           ++ " := let fix nf'ListTree_"
           ++ "        {a0 b0 : Type} `{Normalform Shape Pos a0 b0}"
           ++ "        (x1 : List Shape Pos (Tree Shape Pos a0))"
           ++ "        : Free Shape Pos (List Identity.Shape Identity.Pos"
           ++ "                          (Tree Identity.Shape Identity.Pos b0))"
           ++ "      := match x1 with"
           ++ "         | nil => pure nil"
           ++ "         | cons fx1 fx2 =>"
           ++ "             fx1 >>= (fun x2 => nf'Tree_ x2 >>= (fun nx1 =>"
           ++ "             fx2 >>= (fun x3 => nf'ListTree_ x3 >>= (fun nx2 =>"
           ++ "               pure (cons (pure nx1) (pure nx2))))))"
           ++ "         end"
           ++ "    in match x with"
           ++ "       | leaf fx => nf fx >>= (fun nx => pure (leaf (pure nx)))"
           ++ "       | branch fx0 =>"
           ++ "           fx0 >>= (fun x0 => nf'ListTree_ x0 >>= (fun nx0 =>"
           ++ "             pure (branch (pure nx0))))"
           ++ "       end. "
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  applyBind :: Coq.Term -> Coq.Term -> Coq.Term
+  applyBind mx f = Coq.app (Coq.Qualid Coq.Base.freeBind) [mx, f]
+
+  -- | Given an A, returns Free Shape Pos A
```
You often forgot to use Haddock syntax for code in comments.

```suggestion
  -- | Given an @A@, returns @Free Shape Pos A@.
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+
+  -- | Shape and Pos arguments as Coq terms.
+  shapeAndPos :: [Coq.Term]
+  shapeAndPos = [Coq.Qualid Coq.Base.shape, Coq.Qualid Coq.Base.pos]
```
Use `genericApply` whenever you have to apply `Shape` and `Pos`.

```suggestion
  applyFree :: Coq.Term -> Coq.Term
  applyFree a = genericApply Coq.Base.free [] [] [a]
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+
+  -- | Shortcut for the application of >>=.
+  applyBind :: Coq.Term -> Coq.Term -> Coq.Term
+  applyBind mx f = Coq.app (Coq.Qualid Coq.Base.freeBind) [mx, f]
```
Try to use `generateBind` instead.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+    return (Coq.app l' [r'], varsl ++ varsr)
+  -- Function types were removed by 'stripType'.
+  toCoqType _ _ (IR.FuncType _ _ _)
+    = error "Function types should have been eliminated."
```
You can probably rewrite this function in terms of `insertFreshVariables` and `convertType'`. To get the names of the generated fresh variables, you can use `freeTypeVars`. To rename `Shape` and `Pos` to `Identity.Shape` and `Identity.Pos` you could use the [substitutions for Coq terms](https://github.com/FreeProving/language-coq/blob/main/src/lib/Language/Coq/Subst.hs) that are provided by `language-coq`.

MajaRet replied:
I implemented this suggestion partially. I did not use `convertType'` or `freeTypeVars` as that would, in my opinion, have actually made the definition more complicated than it is now. `insertFreshVariables` also does not insert any variables into the environment, which would have had to be changed to use those functions.
But I simplified the function so it is now only a few lines long and use Coq term substitutions where appropriate.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+      -- Get the Coq name of the constructor.
+      let conIdent = entryIdent conEntry
+      -- Generate fresh variables for the constructor's parameters.
+      conArgIdents <- freshQualids (entryArity conEntry) ("f" ++ freshArgPrefix)
```
Add prefixes such as `fx`, `nx` and `sx` to `FreeC.Environment.Fresh` such that there is a single point where all prefixes are configured.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  -- | A type variable that represents irrelevant parts of a type expression.
+  -- Represented by an underscore.
+  placeholderVar :: IR.Type
+  placeholderVar = IR.TypeVar NoSrcSpan "_"
```
Your current uses of `_` seem to be safe but the `_` could introduce bugs that are difficult to debug in the future. I'd suggest to make explicit on type level whether you are dealing with an actual type or a type with underscores. You can either introduce a wrapper type

```haskell
newtype StrippedType = StrippedType IR.Type
```

which can only be destructed by `insertFreshVariables` or preferably have a special representation such as

```haskell
data StrippedType = StrippedType | StrippedTypeCon IR.TypeConName [StrippedType]
```

which would also make some of the algorithms in this module easier to comprehend.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+-- Instance Generation                                                       --
+-------------------------------------------------------------------------------
+-- | Builds instances for all supported typeclasses.
+--   Currently, only a @Normalform@ instance is generated.
```
`ShareableArgs` instances are generated as well.

```suggestion
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  -- The types of the data declaration's constructors' arguments.
+  let argTypes = map (concatMap IR.conDeclFields . IR.dataDeclCons) dataDecls
+  -- The same types where all type synonyms are expanded.
+  argTypesExpanded
+    <- mapM (mapM expandAllTypeSynonyms) argTypes -- :: [[IR.Type]]
```
```suggestion
  argTypesExpandedn <- mapM (mapM expandAllTypeSynonyms) argTypes
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+  -- in topological order.
+  let reducedTypes = map (nub . reverse . concatMap collectSubTypes)
+        argTypesExpanded
+  -- Like reducedTypes, but with all occurrences of the types for which we are defining
```
Use Haddock syntax in non-documentation comments as well.

```suggestion
  -- Like 'reducedTypes', but with all occurrences of the types for which we are defining
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+
+  -- | Constructs instances of a typeclass for a set of mutually recursive
+  --   types. The typeclass is specified by the arguments.
+  buildInstances
+    ::
+    -- For each data declaration, this list contains the occurrences of
+    -- indirect recursion in the constructors of that data declaration.
+    [[IR.Type]]
+    -> String -- The name of the class function.
+    -> String -- The name of the typeclass.
+    -> (IR.Type -- The type for which the instance is being defined.
+        -> Coq.Qualid -- The name of a variable of that type.
+        -> Converter ([Coq.Binder], Coq.Binder, Coq.Term, Coq.Term))
+    -> (TypeMap -- A mapping from types to function names.
+        -> Coq.Qualid -- The name of a constructor.
+        -> [(IR.Type, Coq.Qualid)]
+        -> Converter Coq.Term)
+    -> Converter [Coq.Sentence]
```
Use `-- ^` for comments of function arguments.

```suggestion
buildInstances
  :: [[IR.Type]]
  -- ^ For each data declaration, this list contains the occurrences of
  --   indirect recursion in the constructors of that data declaration.
  -> String         -- ^ The name of the class function.
  -> String         -- ^ The name of the typeclass.
  -> (IR.Type       -- ^ The type for which the instance is being defined.
      -> Coq.Qualid -- ^ The name of a variable of that type.
      -> Converter ([Coq.Binder], Coq.Binder, Coq.Term, Coq.Term))
  -> (TypeMap       -- ^ A mapping from types to function names.
      -> Coq.Qualid -- ^ The name of a constructor.
      -> [(IR.Type, Coq.Qualid)]
      -> Converter Coq.Term)
  -> Converter [Coq.Sentence]

```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+--   Currently, only a @Normalform@ instance is generated.
+--
+--   Suppose we have a type
+--   > data T a1 ... an = C1 a11 ... a1m1 | ... | Ck ak1 ... akmk.
```
Surround block-level markup with blank lines.

```suggestion
--
--   > data T a1 ... an = C1 a11 ... a1m1 | ... | Ck ak1 ... akmk.
--
```

Consider to use Unicode characters for indices, ellipses, meta-variables etc. like in the documentation of other modules.

```suggestion
--
--   > data T α₁ ... αₙ = C₁ a₍₁,₁₎ ... a₍₁,ₘ₁₎ | … | Cₖ a₍ₖ,₁₎ … a₍ₖ,ₘₖ₎.
--
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs:
```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+    shareableArgsClassName shareArgsBindersAndReturnType buildShareArgsValue
+  return (nfInstances ++ shareableArgsInstances)
+ where
+  -- The (mutually recursive) data types for which we are defining
+  -- instances, converted to types.
+  declTypes :: [IR.Type]
+  declTypes = map dataDeclToType dataDecls
```
Since `dataDeclToType` is used nowhere else, you could inline it here, for example, by using list comprehensions. That type variables are replaced by `_` may also be worth mentioning in the comment.

```suggestion
-- The (mutually recursive) data types for which we are defining
-- instances, converted to types.
declTypes :: [IR.Type]
declTypes = [IR.typeConApp NoSrcSpan (IR.typeDeclQName dataDecl)
              (replicate (length (IR.typeDeclArgs dataDecl)) placeholderVar)
            | dataDecl <- dataDecls
            ]
```

just95 reviewed (changes requested):
The module for the conversion of type declarations is getting quite large now. We should consider to move the generation of induction schemes, type class instances and smart constructors into their own modules.

Commit:
Apply suggestions #150

Commit:
Fix HLint and Haddock errors #150

Commit:
Move instance and induction scheme generation into separate modules #150

Commit:
Format code #150

Commit:
Test unformatted as-pattern #150

Commit:
Try workaround without as-pattern #150

Commit:
Rename source span variables #150

Commit:
Also rename occurrences of the source span variables #150

Commit:
Change Floskell configuration so as-patterns are not split #150

Commit:
Capitalize words in header comments #150

Commit:
Merge branch 'main' of https://github.com/FreeProving/free-compiler into issue-150

Commit:
Format TypeDecl tests #150

Commit:
Add prefixes for normalized and shared variables to Fresh #150

Commit:
Use new data type for stripped type #150

Commit:
Remove Ord instance from SrcSpan and Type #150

Commit:
Incorporate changes from issue-202 pertaining to Normalform generation #150

Commit:
Adjust TypeDeclTests #150

Commit:
Merge main into branch #150

Commit:
Remove second type variable from Handlers #150

Commit:
Use generateBind #150

Commit:
Remove redundant brackets and fix var names in tests #150

MajaRet commented:
I have implemented the last few suggestions now, but I have also incorporated a few changes into this pull request that were originally required by #202. However, implementing some of the suggestions on what would have essentially been an outdated version of the instance generation would not have made sense. Therefore, those changes have now been added here.

Specifically, the `Normalform` typeclass is now no longer parameterized over two types (plus `Shape` and `Pos`), but only one. So the `Normalform` class itself, the `Prelude` instances, and the `Normalform`-specific functions in the instance generation function have changed to reflect that.
